mirrord/agent/src/steal/connection.rs

```diff
@@ -591,28 +604,131 @@ impl TcpConnectionStealer {
+            Version::HTTP_10 => "HTTP/1.0",
+            Version::HTTP_11 => "HTTP/1.1",
+            Version::HTTP_2 => "HTTP/2",
+            Version::HTTP_3 => "HTTP/3",
+            _ => "HTTP/1.1", // TODO: What should happen here?
+        };
+
+        // Status line.
+        bytes.put(version.as_bytes()); // HTTP/1.1
+        bytes.put(&b" "[..]);
+        bytes.put(status.as_str().as_bytes()); // 200
+        bytes.put(&b" "[..]);
+        if let Some(reason) = status.canonical_reason() {
+            bytes.put(reason.as_bytes()); // OK
+        } else {
+            bytes.put(&b"LOL"[..]);
+        }
+
+        // Headers.
+        for (name, value) in headers.into_iter() {
+            if let Some(name) = name {
+                bytes.put(&b"\r\n"[..]);
+                bytes.put(name.as_str().as_bytes());
+                bytes.put(&b": "[..]);
+            } else {
+                // TODO: are multiple values for the same header comma delimited?
+                bytes.put(&b","[..])
+            }
+            bytes.put(value.as_bytes())
+        }
+        bytes.put(&b"\r\n\r\n"[..]);
+
+        // Body.
+        bytes.put(body.collect().await.unwrap().to_bytes()); // Unwrap Infallible.
+
+        bytes.to_vec()
+    }
+
+    fn handle_response_reconstruction_fail(err: hyper::http::Error) -> hyper::http::Error {
+        error!("Could not reconstruct http response: {err:?}");
+        // TODO: What should happen? Currently just skipping that response, which would
+        //       also spoils the rest of the stream. Should hopefully never happen.
+        debug_assert!(false);
+        err
+    }
+
+    /// Forward an HTTP response to a stolen HTTP request from the layer to the back to the browser.
     ///
-    ///                       _________________________agent_________________________
-    /// Local App -> Layer -> ClientConnectionHandler -> Stealer -> HttpFilterManager -> Browser
-    ///                                                          ^- You are here.
-    async fn http_response(
-        &mut self,
-        response: HttpResponse,
-    ) -> std::result::Result<(), AgentError> {
-        todo!();
-        match self.port_subscriptions.get(&response.port) {
-            Some(HttpFiltered(manager)) => todo!(),
-            Some(Unfiltered(_)) | None => {
-                warn!(
-                    "Got an http response to port {:?} which is not being filtered. \
-                Maybe the layer unsubscribed and this is a remainder of a closed connection. \
-                Not forwarding.",
-                    response.port
-                );
-                Ok(())
+    ///                         _______________agent_______________
+    /// Local App --> Layer --> ClientConnectionHandler --> Stealer --> Browser
+    ///                                                             ^- You are here.
+    async fn http_response(&mut self, response: HttpResponse) -> Result<()> {
+        let tcp_stream_write_half;
```

Comment:
Pretty weird name plus you're retrieving it from a hash called `http_write_streams`?

Reply:
Is `http_tx` better? https://github.com/t4lz/mirrord/pull/3/commits/5642fda59668475f3ddca338ff5c35981e4b655c

Reply:
Yup